ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/AdminNodeInfo.java

```diff
@@ -0,0 +1,113 @@
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcErrorResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcResponse;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import tech.pegasys.pantheon.ethereum.p2p.P2pDisabledException;
+import tech.pegasys.pantheon.ethereum.p2p.api.P2PNetwork;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.net.InetAddresses;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class AdminNodeInfo implements JsonRpcMethod {
+  private static final Logger LOG = LogManager.getLogger();
+  private final String clientVersion;
+  private final int networkId;
+  private final JsonNode genesisConfigConfig;
+  private final P2PNetwork peerNetwork;
+  private final BlockchainQueries blockchainQueries;
+
+  public AdminNodeInfo(
+      final String clientVersion,
+      final int networkId,
+      final JsonNode genesisConfigConfig,
+      final P2PNetwork peerNetwork,
+      final BlockchainQueries blockchainQueries) {
+    this.peerNetwork = peerNetwork;
+    this.clientVersion = clientVersion;
+    this.genesisConfigConfig = genesisConfigConfig;
+    this.blockchainQueries = blockchainQueries;
+    this.networkId = networkId;
+  }
+
+  @Override
+  public String getName() {
+    return "admin_nodeInfo";
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequest req) {
+
+    try {
+      final Map<String, Object> response = new HashMap<>();
+      final BytesValue nodeId = peerNetwork.getLocalPeerInfo().getNodeId();
+      final InetSocketAddress address = peerNetwork.getDiscoverySocketAddress();
+      final int port = peerNetwork.getLocalPeerInfo().getPort();
+
+      final InetAddress inetAddress = address.getAddress();
+      response.put(
+          "enode",
+          "enode://"
+              + nodeId.toString().substring(2)
+              + "@"
+              + InetAddresses.toUriString(inetAddress)
+              + ":"
+              + port);
+      response.put("id", nodeId.toString().substring(2));
+      // this doesn't provide a useful value yet.
+      // response.put("ip", inetAddress.getHostAddress());
+      response.put("listenAddr", InetAddresses.toUriString(inetAddress) + ":" + port);
+      response.put("name", clientVersion);
+      response.put("ports", ImmutableMap.of("discovery", port, "listener", port /*??*/));
+
+      final ChainHead chainHead = blockchainQueries.getBlockchain().getChainHead();
+      response.put(
+          "protocols",
+          ImmutableMap.of(
+              "eth",
+              ImmutableMap.of(
+                  "config",
+                  genesisConfigConfig,
```

Comment:
Not entirely sure what the right answer is here but I don't particularly like passing the "raw" json around so would be nice to use either the parsed `GenesisConfigFile` or `GenesisConfigOptions` here.  That would then also ensure we return Pantheon's understanding of the genesis config, not just the raw text which avoids confusion when some setting isn't taking effect because it's misspelt as it wouldn't appear here at all. 

I'm not sure if that justifies the extra complexity of having to reserialize the JSON or if there would be some reason to want to have unsupported options funnelled through to here.

Reply:
The main question is do we want the literal config elements in the json text or do we want Pantheon's understanding of the options?  Since we are doing live nodeInfo I think we want the understanding.  Wiring up json de/serialization for the relevant interfaces should do the trick.

Reply:
Another option is to ask JsonGenesisConfigOptions for its configRoot, which is already the type we need.  But that is the unparsed and uninterpreted values.

Reply:
Yeah I was wondering that.  The big downside is that it's not quite the raw values (we normalise all keys to lower case) and not just what pantheon understands either.

Reply:
Roger.  I think I can add an interface method to PantheonController to return the GenesisConfigOptions.